lib/searchable.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020-2022 CYBERCRYPT
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encryptonize
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"encryptonize/crypto"
+)
+
+type Index struct {
+	Mapping map[string]SealedID
+}
+
+type SealedID struct {
+	Ciphertext []byte
+	WrappedKey []byte
+}
+
+func NewSearchable() *Index {
+	mapping := make(map[string]SealedID)
+
+	return &Index{Mapping: mapping}
+}
+
+func (i *Index) Add(masterKey []byte, keyword, id string) error {
+	k1 := crypto.KMACKDF(len(masterKey), masterKey, []byte("mac"), []byte(keyword))
+	k2 := crypto.KMACKDF(len(masterKey), masterKey, []byte("id"), []byte(keyword))
+
+	tagger, err := crypto.NewKMAC256Tagger(k1)
+	if err != nil {
+		return err
+	}
+	cryptor, err := crypto.NewAESCryptor(k2)
+	if err != nil {
+		return err
+	}
+
+	count, err := i.Count(masterKey, keyword)
+	if err != nil {
+		return err
+	}
+	label, err := tagger.Tag(uint64Encode(count))
+	if err != nil {
+		return err
+	}
+
+	wrappedKey, encryptedID, err := cryptor.Encrypt([]byte(id), 0)
+	if err != nil {
+		return err
+	}
+	sealedID := SealedID{Ciphertext: encryptedID, WrappedKey: wrappedKey}
+
+	i.Mapping[base64.StdEncoding.EncodeToString(label)] = sealedID
+
+	return nil
+}
+
+func (i *Index) Search(masterKey []byte, keyword string) ([]string, error) {
+	k1 := crypto.KMACKDF(len(masterKey), masterKey, []byte("mac"), []byte(keyword))
+	k2 := crypto.KMACKDF(len(masterKey), masterKey, []byte("id"), []byte(keyword))
+
+	tagger, err := crypto.NewKMAC256Tagger(k1)
+	if err != nil {
+		return nil, err
+	}
+	cryptor, err := crypto.NewAESCryptor(k2)
+	if err != nil {
+		return nil, err
+	}
+
+	decryptedIDs := []string{}
+
+	var count uint64
+	for {
```

Comment:
You can have a for-loop with an empty stop condition:
```suggestion
	for count := 0 ; ; count++ {
```